### PR TITLE
Add Transaction

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This adds the Transaction API to the Lite SDK. It uses the same patterns as https://github.com/firebase/firebase-js-sdk/pull/3152

### CODE REVIEW

File: packages/firestore/src/util/promise.ts [hunk 1]
```diff
@@ -23,7 +23,7 @@ export interface Rejecter {
   (reason?: Error): void;
 }
 
-export class Deferred<R> {
+export class Deferred<R = void> {
```
Comment (thebrianchen) [inline packages/firestore/src/util/promise.ts:26]: What is the purpose of adding `void` to the generic? It seems to compile and run fine even when I removed it.

File: packages/firestore/lite/src/api/transaction.ts [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+import { tryCast } from './util';
+
+export class Transaction implements firestore.Transaction {
+  // This is the lite version of the Transaction API used in the legacy SDK. The
+  // class is a close copy but takes different input types.
+
+  private _dataReader: UserDataReader;
+
+  constructor(
+    private _firestore: Firestore,
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/lite/src/api/transaction.ts]: Done

File: packages/firestore/lite/src/api/transaction.ts [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    const ref = validateReference(documentRef, this._firestore);
+    return this._transaction
+      .lookup([ref._key])
+      .then((docs: MaybeDocument[]) => {
+        if (!docs || docs.length !== 1) {
+          return fail('Mismatch in docs returned from document lookup.');
+        }
+        const doc = docs[0];
+        if (doc instanceof NoDocument) {
+          return new DocumentSnapshot<T>(this._firestore, ref._key, null);
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/lite/src/api/transaction.ts]: Yes! Thanks for catching.

File: packages/firestore/src/util/promise.ts (hunk 1 above)
Comment (schmidt-sebastian) [reply packages/firestore/src/util/promise.ts:26]: Without it `new Deferred()` is `Deferred<unknown>`. I would like it to default it to `Deferred<void>`.

**Pull request merged**